### PR TITLE
[CSL-744] Haskeline-based REPL

### PR DESCRIPTION
**Pull request by int-index**
it's ready for review, but I'm going to add some documentation a bit later (perhaps tomorrow)

The general idea is that Haskeline UI is running in a separate thread and communicates with `auxxPlugin` via a couple of `MVar`s. The logic of communication is neatly encapsulated in `createAuxxRepl` and the rest of the code doesn't need to deal `MVar`s.

### CODE REVIEW

File: auxx/src/Plugin.hs [hunk 1]
```diff
@@ -43,52 +44,30 @@ import           Mode                       (AuxxMode)
...
+    -> Either (WithCommandAction AuxxMode) Text
     -> (WorkerSpec AuxxMode, OutSpecs)
-auxxPlugin AuxxOptions {..} =
-    case aoAction of
-        Repl    -> worker' runCmdOuts $ runWalletRepl
-        Cmd cmd -> worker' runCmdOuts $ runWalletCmd cmd
+auxxPlugin AuxxOptions{..} = \case
+        Left WithCommandAction{..} -> worker' runCmdOuts $ \sendActions -> do
+            printAction <- getPrintAction
+            printAction "... the auxx plugin is ready"
```
Comment (int-index) [inline auxx/src/Plugin.hs]: This message doesn't disrupt the REPL if the user started entering a command before the `auxxPlugin` is ready to handle commands :tada:

File: auxx/src/Main.hs [hunk 2]
```diff
@@ -104,5 +105,10 @@ action opts@AuxxOptions {..} = withConfigurations conf $ do
 main :: IO ()
 main = withCompileInfo $(retrieveCompileTimeInfo) $ do
     opts <- getAuxxOptions
-    let loggingParams = CLI.loggingParams "auxx" (aoCommonNodeArgs opts)
-    loggerBracket loggingParams $ runProduction $ action opts
+    let loggingParams = (CLI.loggingParams "auxx" (aoCommonNodeArgs opts))
+                        { lpConsoleLog = Just False }
```
Comment (gromakovsky) [inline auxx/src/Main.hs]: I think console log should be disabled only for repl, am I wrong?

File: auxx/src/Repl.hs [hunk 3]
```diff
@@ -0,0 +1,174 @@
...
+        --
+    }
+
+-- | The action to post a command to be executed by the auxx plugin and wait for
+-- its result. This action blocks until the result is available: that is, until
+-- command execution completes or an exception is thrown.
+type PutCommandAction = Command -> AuxxRepl CommandResult
+
+-- | Create the auxx REPL (as an abstract @'IO' ()@ action) and the interface to
+-- interact with it ('WithcommandAction'). You are supposed to fork the REPL
```
Comment (gromakovsky) [inline auxx/src/Repl.hs]: `WithCommandAction` (typo)

File: auxx/src/Repl.hs [hunk 4]
```diff
@@ -0,0 +1,174 @@
...
+auxxReplOn printActionVar putCommand = Haskeline.runInputT settings $ do
+    printAction <- Haskeline.getExternalPrint
+    putMVar printActionVar (liftIO . printAction . toString)
+    greetings
+    loop (prompt putCommand)
+  where
+    greetings = do
+        outputTextLn "Welcome to Cardano SL Auxx REPL"
+        outputTextLn helpMessage
+    settings = Haskeline.Settings Haskeline.noCompletion Nothing True
```
Comment (gromakovsky) [inline auxx/src/Repl.hs]: Why not use explicit records names?

File: auxx/src/Main.hs (hunk 2 above)
Comment (int-index) [reply auxx/src/Main.hs]: The node tries to print to `stdout`, it disrupts the REPL.

File: auxx/src/Repl.hs (hunk 3 above)
Comment (int-index) [reply auxx/src/Repl.hs]: thanks, fixed

File: auxx/src/Repl.hs (hunk 4 above)
Comment (int-index) [reply auxx/src/Repl.hs]: good idea, fixed

File: auxx/src/Repl.hs [hunk 5]
```diff
@@ -0,0 +1,178 @@
...
+    putMVar printActionVar (liftIO . printAction . toString)
+    greetings
+    loop (prompt putCommand)
+  where
+    greetings = do
+        outputTextLn "Welcome to Cardano SL Auxx REPL"
+        outputTextLn helpMessage
+    settings = Haskeline.Settings
+        { Haskeline.complete       = Haskeline.noCompletion
+        , Haskeline.historyFile    = Nothing -- TODO: pass it from CLI
```
Comment (gromakovsky) [inline auxx/src/Repl.hs]: I created https://issues.serokell.io/issue/CSL-1779
We can do it later, but can you add issue id to this TODO?

File: auxx/src/Main.hs (hunk 2 above)
Comment (gromakovsky) [reply auxx/src/Main.hs]: You probably got me wrong, I mean that if `aoAction opts` is `Cmd`, do we want to disable console output?

File: auxx/src/Repl.hs (hunk 5 above)
Comment (int-index) [reply auxx/src/Repl.hs]: done

File: auxx/src/Main.hs (hunk 2 above)
Comment (int-index) [reply auxx/src/Main.hs]: ah, I see. Alright, fixed.

**Pull request merged**